presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPlanRemoteProjections.java

```diff
@@ -0,0 +1,264 @@
+import static com.facebook.presto.sql.planner.iterative.rule.PlanRemotePojections.ProjectionContext;
+import static org.testng.Assert.assertEquals;
+
+public class TestPlanRemoteProjections
+        extends BaseRuleTest
+{
+    public static final QualifiedFunctionName REMOTE_FOO = QualifiedFunctionName.of(new CatalogSchemaName("unittest", "memory"), "remote_foo");
+    public static final RoutineCharacteristics.Language JAVA = new RoutineCharacteristics.Language("java");
+
+    public static final SqlInvokedFunction FUNCTION_REMOTE_FOO_0 = new SqlInvokedFunction(
+            REMOTE_FOO,
+            ImmutableList.of(),
+            parseTypeSignature(StandardTypes.INTEGER),
+            "remote_foo()",
+            RoutineCharacteristics.builder().setLanguage(JAVA).setNullCallClause(RETURNS_NULL_ON_NULL_INPUT).build(),
+            "",
+            Optional.empty());
+
+    public static final SqlInvokedFunction FUNCTION_REMOTE_FOO_1 = new SqlInvokedFunction(
+            REMOTE_FOO,
+            ImmutableList.of(new Parameter("x", parseTypeSignature(StandardTypes.INTEGER))),
+            parseTypeSignature(StandardTypes.INTEGER),
+            "remote_foo(x)",
+            RoutineCharacteristics.builder().setLanguage(JAVA).setDeterminism(DETERMINISTIC).setNullCallClause(RETURNS_NULL_ON_NULL_INPUT).build(),
+            "",
+            Optional.empty());
+
+    public static final SqlInvokedFunction FUNCTION_REMOTE_FOO_2 = new SqlInvokedFunction(
+            REMOTE_FOO,
+            ImmutableList.of(new Parameter("x", parseTypeSignature(StandardTypes.INTEGER)), new Parameter("y", parseTypeSignature(StandardTypes.INTEGER))),
+            parseTypeSignature(StandardTypes.INTEGER),
+            "remote_foo(x, y)",
+            RoutineCharacteristics.builder().setLanguage(JAVA).setDeterminism(DETERMINISTIC).setNullCallClause(RETURNS_NULL_ON_NULL_INPUT).build(),
+            "",
+            Optional.empty());
+
+    public static final SqlInvokedFunction FUNCTION_REMOTE_FOO_3 = new SqlInvokedFunction(
+            REMOTE_FOO,
+            ImmutableList.of(new Parameter("x", parseTypeSignature(StandardTypes.INTEGER)), new Parameter("y", parseTypeSignature(StandardTypes.INTEGER)), new Parameter("z", parseTypeSignature(DOUBLE))),
+            parseTypeSignature(StandardTypes.INTEGER),
+            "remote_foo(x, y, z)",
+            RoutineCharacteristics.builder().setLanguage(JAVA).setNullCallClause(RETURNS_NULL_ON_NULL_INPUT).build(),
+            "",
+            Optional.empty());
+
+    private static final Metadata METADATA = MetadataManager.createTestMetadataManager();
+
+    public TestPlanRemoteProjections()
+    {
+        FunctionManager functionManager = METADATA.getFunctionManager();
+        functionManager.addTestFunctionNamespace("unittest", new InMemoryFunctionNamespaceManager("unittest", new SqlInvokedFunctionNamespaceManagerConfig().setSupportedFunctionLanguages("sql,java")));
```

Comment:
Let's use H2 FNM, which is designed for testing purpose. We can then avoid adding test-related method `addTestFunctionNamespace` in our production code.

You can do that by first adding the factory:
```Java
static {
    METADATA.getFunctionManager().addFunctionNamespaceFactory(new H2FunctionNamespaceManagerFactory());
}
```

You'll then be able to add a function namespace by
```Java
functionManager.loadFunctionNamespaceManager(
    H2FunctionNamespaceManagerFactory.NAME,
    "unittest",
    ImmutableMap.of(<config key>, <config value>))
```

Reply:
I remember I tried that but it didn't work. But I don't remember what didn't work. This is the problem of delayed review... Let me try again I guess...

Reply:
I think the reason to not use `H2FunctionNamespaceManagerFactory` is dependencies. This test is in presto-main. `H2FunctionNamespaceManagerFactory` is in presto-tests.